### PR TITLE
[Large] Create UI for metadata editing workflow

### PR DESCRIPTION
**Pull request by aswallace**
This PR creates the main part of the UI for editing file metadata from BFF based on [the Figma designs](https://www.figma.com/design/oNQvOAK11YINcbPvNmkgwO/Public-File-Explorer?node-id=875-4724&node-type=canvas&t=JUN2ieb9MDJMnUxO-0). Does not include any backend logic.

~[Temporarily deployed to staging](https://staging.biofile-finder.allencell.org/app)~

Part of #96 

## Reviewing
Lots of changes with a slightly complicated file structure, so here's suggested review order:
- Entrypoint: added edit to the dropdown menu in [packages/core/hooks/useFileAccessContextMenu.ts](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-71bcb3b6133f882ebdd1dd66c282d671a299b33032af1c33744369ec05c99339)
- This opens the modal, which is [packages/core/components/Modal/EditMetadata/index.tsx](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-a6ec79eb3ac34e5ecd40a3aead96f6e047d6d3143b78100b3d65728380350a9f)
- The modal renders the actual editing component ([packages/core/components/EditMetadata/index.tsx](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-264e2d9d82ffe4d96144ff07cb2dac39ff71aab629579b15dbf783ce1a68d9fc)), which is a wrapper for both the "new" and "existing" annotation pathways 
    - Pathway for editing files' values for an existing annotation: [packages/core/components/EditMetadata/ExistingAnnotationPathway.tsx](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-f4052c195644362a2061021276d6bb4298a6a4d91c4cedd54acd4b25a461a490)
    - Pathway for creating a new annotation on the files: [packages/core/components/EditMetadata/NewAnnotationPathway.tsx](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-a6c315bc57c9c1a4506e78753a90b1b1a4445bd9f28a80ccdca3e358b3f91b25)
- BOTH pathways then point at [packages/core/components/EditMetadata/MetadataDetails.tsx](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-b43d09942ca733e4d46c2cdaf4192858058d787863858b8195a302c7c055e9b8), which displays the current values for the files and provides an input field to enter the desired replacement values

There's also a brand new ['ComboBox' component](https://github.com/AllenInstitute/biofile-finder/pull/338/files#diff-ecbdd5e8e843d5406ce8ffcc340ebe5ac1035cf11ba43ed7aca8929eac8e1f01) (dropdown selector with search), which is a styled wrapper for the [FluentUI combo box](https://developer.microsoft.com/en-us/fluentui#/controls/web/combobox)

## To Dos (UI only)
- Unit tests
- #340 (date, number, yes/no, etc)
- #341
- #342 -- Ideally should open on click and react to users hitting enter, but has been behaving inconsistently
- #343 Add disabled annotations (e.g., from labkey) & prevent new annotations that collide with existing
- #344 Warn users on exit w/ unsaved changes

### Existing annotation pathway: 
Screenshot 2024-11-19 at 4 42 56 PM
Screenshot 2024-11-19 at 4 43 10 PM
Screenshot 2024-11-19 at 4 43 33 PM

### New annotation pathway:
Screenshot 2024-11-19 at 4 43 51 PM
Screenshot 2024-11-19 at 4 44 09 PM
Screenshot 2024-11-19 at 4 44 16 PM

### CODE REVIEW

**Review by BrianWhitneyAI — APPROVED** [review]

**Review by kmitcham — APPROVED** [review]

File: packages/core/components/EditMetadata/index.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    },
+                ]}
+            />
+            <div className={styles.contentWrapper}>
+                {editPathway === EditMetadataPathway.EXISTING ? (
+                    <ExistingAnnotationPathway
+                        onDismiss={onDismiss}
+                        annotationValueMap={annotationValueMap}
+                        annotationOptions={annotationOptions}
+                        selectedFileCount={fileSelection.count()}
```
[the remark this answers is not in the document]
Comment (aswallace) [reply packages/core/components/EditMetadata/index.tsx]: [86841dd](https://github.com/AllenInstitute/biofile-finder/pull/338/commits/86841ddb75322e2260230cf5c2422d5fa987ecf3)

**Comment by aswallace** [comment]
UX bugs from @lynwilhelm : 
- [ ] Arrow down with menu open changes file selection in the background (yikes!)
- [ ] Don't allow user to move modal & fix move cursor
- [ ] Remove REPLACE from first step
- [ ] Table shouldn't be focusable
- [ ] Fix spacing for "Existing" header
- [ ] Fix table header spacing & font size
- [ ] Add tooltips on radio options (new/existing)
- [ ] Tab sometimes advances the modal when it shouldn't
- Combobox menu bugs (may need to convert these to separate issue)
    - [ ] When text is entered in annotation combobox, clicking the 'x' goes to the next step instead of dismissing
    - [ ] Weird text highlighting & case restriction in autocomplete
    - [ ] Add overlay to menu overflow & increase height slightly 
    - [ ] Add styling in combobox callout for focus/active/selected
    - [ ] Clicking inside & typing still doesn't open open menu by default
    - [ ] Searched item does not always appear at the top of the search results

File: packages/core/hooks/useFileAccessContextMenu.ts [hunk 2]
```diff
@@ -135,6 +136,24 @@ export default (filters?: FileFilter[], onDismiss?: () => void) => {
                         ],
                     },
                 },
+                ...(isQueryingAicsFms
```
Comment (SeanLeRoy) [inline packages/core/hooks/useFileAccessContextMenu.ts:139]: Did you create a ticket for adding this capability for non-aics data sources? I think it wouldn't be very difficult to do via SQL like:
```
UPDATE ${datasourceName}
SET ${columnToUpdate} = ${valueToReplaceWith}
WHERE ${here either select by row number or the internally made BFF ID} = ${list of row numbers or bff ids}
```

File: packages/core/components/EditMetadata/index.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ */
+export default function EditMetadataForm() {
+    const dispatch = useDispatch();
+    const onDismiss = () => {
+        dispatch(interaction.actions.hideVisibleModal());
+    };
+    const fileSelection = useSelector(selection.selectors.getFileSelection);
+    const fileCount = fileSelection.count();
+    // Don't allow users to edit top level annotations (e.g., File Name)
+    const annotationOptions = useSelector(metadata.selectors.getSortedAnnotations)
```
Comment (SeanLeRoy) [inline packages/core/components/EditMetadata/index.tsx:29]: Unsure why I have never attached JS array functions onto the selector like this... looks clean!

**Review by SeanLeRoy — APPROVED** [review]
Anya! You nailed this 💯

File: packages/core/hooks/useFileAccessContextMenu.ts (hunk 2 above)
Comment (aswallace) [reply packages/core/hooks/useFileAccessContextMenu.ts:139]: honestly I forgot we were planning to allow editing on non-aics data sources. Will make a ticket!

File: packages/core/components/EditMetadata/ExistingAnnotationPathway.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    ) => {
+        let valueMap: ValueCountItem[] = [];
+        // FluentUI's combobox doesn't always register the entered value as an option,
+        // so we need to be able to check both
+        const selectedFieldName = option?.text || value;
+        if (!selectedFieldName) return;
+        // Track how many values we've seen, since some files may not have a value for this field
+        let totalValueCount = 0;
+        if (props?.annotationValueMap?.has(selectedFieldName)) {
+            const fieldValues = props.annotationValueMap.get(selectedFieldName);
```
[the remark this answers is not in the document]
Comment (aswallace) [reply packages/core/components/EditMetadata/ExistingAnnotationPathway.tsx]: Changed in [6e4de6a](https://github.com/AllenInstitute/biofile-finder/pull/338/commits/6e4de6a528c139e8afd4d861191b9dfe9f0d4876)

**Pull request merged**